### ISSUE TITLE
Adapt zos4 storage to host the ipxe bootblock on Disk

### ISSUE THREAD

**Issue by delandtj** [threefoldtech/zos#2506]
## Why

To be able to have zos...4 boot in Hetzner, we need to have the ipxe bootblock (or UEFI boot) on the bootable disk, because booting over USB in Hetzner requires a whole set of manual intervention and request for Hands-on actions and tickets. While cumbersome, it's also **very** time-consuming.

## What 
At Hetzner you can use an install script (using their rescue system, which is basically an ipxe and nfsroot system) to install a system.
We can leverage this to 'just' install the iPXE image (img or UEFI) on a small partition on the bootdisk of the system.
To make things as small as possible in zos, we are going to create A GPT labeled boot partition of a few MiB, create the second partition for the rest of the disk, taking care of correct alignment, where the boot partition is an MSDOS VFAT fs with LABEL ZOSPXE, and the second partition will just have a btrfs type.

Zos then needs to detect the ZOSPXE label, and if it exists, not partition the disk, but use the second partition as-is for creating a btrfs filesystem on it and carry on with the other disks as normal (i.e. create an FS on it when empty or don't touch when existing partitions/filesystems).

**Comment by delandtj** [comment]
@ashraffouda : if questions, ask .

**Comment by muhamadazmy** [comment]
Just a small note to be considered:
Zos only uses disks that has no partitions. It's always required to have clean disks with no partition table what so ever.

If we now start creating EFI partition to copy the boot usb stick, and when ZOS is booting it will complain that there are no usable disks available.

It's then might be better if zos4 can actually install itself on the disk, maybe ?

**Comment by delandtj** [comment]
no install , just a bootblock
and it's not that difficult is it ? 
disk not empty -> is there a partition with a vfat label ZOSPXE? aight mount second partition. that's all.

**Comment by muhamadazmy** [comment]
No, it's still possible of course. Some changes to storaged to allow using disks with partitions or special handling of the EFI partition of the cache disk

**Comment by delandtj** [comment]
has something happened ?

**Comment by xmonader** [comment]
notes: 

- can use blkid to detect ZOSPXE (can be in disk.go)
- in storage.go/initialize skip the zospxe device from the btrfs pool

@muhamadazmy can you confirm that's the right approach?

**Comment by muhamadazmy** [comment]
I am wondering if special handling is needed at all. Maybe instead of requiring a completely empty disk we allow creating new partitions on gpt disks and mount them instead of a full disk. I think this will make the code easier to maintain.
Zos will still need to support full disks of course for backward compatibility.
This way the setup script just can create the gpt table, and the boot EFI partition, then the system when booted it will just create a partition on the remaining empty space.

**Comment by delandtj** [comment]
> I am wondering if special handling is needed at all. Maybe instead of requiring a completely empty disk we allow creating new partitions on gpt disks and mount them instead of a full disk. I think this will make the code easier to maintain. Zos will still need to support full disks of course for backward compatibility. This way the setup script just can create the gpt table, and the boot EFI partition, then the system when booted it will just create a partition on the remaining empty space.

yeah well... In a later iteration maybe
